hugr-passes/src/static_circuit.rs

```diff
@@ -0,0 +1,398 @@
+struct QbHistory(Node, Vec<Gate>);
+
+impl QbHistory {
+    pub fn init(qb: Node) -> Self {
+        Self(qb, vec![])
+    }
+
+    pub fn push(&mut self, node: Node, g: impl Into<String>, qbs: impl IntoIterator<Item = Node>) {
+        self.1.push(Gate::new(node, g.into(), qbs))
+    }
+
+    fn qb_index(&self) -> Node {
+        self.0
+    }
+}
+
+impl TryFrom<Sum<QbHistory>> for QbHistory {
+    type Error = NoDefaultConversionToSum<QbHistory>;
+
+    fn try_from(_: Sum<QbHistory>) -> Result<Self, Self::Error> {
+        Err(NoDefaultConversionToSum::default())
+    }
+}
+
+impl AbstractValue for QbHistory {
+    fn try_join(self, other: Self) -> Option<Self> {
+        eprintln!("join: {:?} ^ {:?} ", &self, &other);
+        let r = if self.0 == other.0 && self.1.len() == other.1.len() {
+            Some(QbHistory(
+                self.0,
+                zip_eq(self.1, other.1)
+                    .map(|(a, b)| a.join(b))
+                    .collect::<Option<Vec<_>>>()?,
+            ))
+        } else {
+            None
+        };
+        panic!("dougrulz");
+        eprintln!("join:  => {:?}", &r);
+        r
+    }
+
+    fn try_meet(self, other: Self) -> Option<Self> {
+        (self.0 == other.0).then_some(self)
+    }
+}
+
+pub struct StaticCircuitContext<'a, H> {
+    pub hugr: &'a H,
+    pub extensions: ExtensionSet,
+    pub alloc_ops: HashSet<Node>,
+    pub free_ops: HashSet<Node>,
+}
+
+impl<H> Clone for StaticCircuitContext<'_, H> {
+    fn clone(&self) -> Self {
+        Self {
+            hugr: self.hugr,
+            extensions: self.extensions.clone(),
+            alloc_ops: self.alloc_ops.clone(),
+            free_ops: self.free_ops.clone(),
+        }
+    }
+}
+
+impl<'a, H: HugrView> StaticCircuitContext<'a, H> {
+    fn assert_invariants(&self) {
+        debug_assert!(self.alloc_ops.is_disjoint(&self.free_ops));
+        for &n in self.alloc_ops.iter().chain(&self.free_ops) {
+            debug_assert!(self.hugr.valid_node(n));
+            let Some(e) = self.hugr.get_optype(n).as_extension_op() else {
+                panic!("node {n} is not in extensions")
+            };
+            debug_assert!(self.extensions.contains(e.def().extension()));
+        }
+    }
+}
+
+impl<'a, H: HugrView> ConstLoader<QbHistory> for StaticCircuitContext<'a, H> {}
+
+impl<'a, H> Deref for StaticCircuitContext<'a, H> {
+    type Target = H;
+
+    fn deref(&self) -> &Self::Target {
+        self.hugr
+    }
+}
+
+impl<'a, H: HugrView> DFContext<QbHistory> for StaticCircuitContext<'a, H> {
+    type View = H;
+
+    fn interpret_leaf_op(
+        &self,
+        node: Node,
+        e: &ExtensionOp,
+        ins: &[PartialValue<QbHistory>],
+        outs: &mut [PartialValue<QbHistory>],
+    ) {
+        self.assert_invariants();
+        let extension = e.def().extension();
+        if !self.extensions.contains(extension) {
+            return;
+        }
+
+        if ins.iter().any(|x| x == &PartialValue::Bottom) {
```

Comment:
I think this demonstrates exactly why I put in the test [you comment on here](https://github.com/CQCL/hugr/pull/1476/#discussion_r1843963953) !!